lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -2536,70 +2551,135 @@ bool EliminateSliceConcat::run(Function *F, const CompilationContext &cctx) {
+    // after the concat dimension.
+    std::vector<
+        std::pair<int /* dimension of slicing */, std::vector<SliceNode *>>>
+        consecutiveSlices;
     std::vector<SliceNode *> currConsecutiveSlices;
     SliceNode *lastSN = nullptr;
+    int lastDim = -1;
     for (auto &concatInput : CN->getInputs()) {
       auto *SN = dyn_cast<SliceNode>(concatInput.getNode());
       // slices with multiple users will not be considered
       if (!SN || SN->getResult().getNumUsers() > 1) {
         if (currConsecutiveSlices.size()) {
-          consecutiveSlices.emplace_back(currConsecutiveSlices);
+          consecutiveSlices.emplace_back(lastDim, currConsecutiveSlices);
           currConsecutiveSlices.clear();
         }
+        lastDim = -1;
         lastSN = nullptr;
         continue;
       }
       // slices with different sources will not be considered
+      unsigned_t dim = -1;
       if (lastSN && (lastSN->getInput() != SN->getInput() ||
-                     !areSlicesConsecutive(lastSN, SN, CN->getDim()))) {
-        consecutiveSlices.emplace_back(currConsecutiveSlices);
+                     !findConsecutiveSliceDim(lastSN, SN, &dim))) {
+        consecutiveSlices.emplace_back(lastDim, currConsecutiveSlices);
         currConsecutiveSlices.clear();
       }
+      lastDim = dim;
       lastSN = SN;
       currConsecutiveSlices.emplace_back(SN);
     }
     if (currConsecutiveSlices.size()) {
-      consecutiveSlices.emplace_back(currConsecutiveSlices);
+      consecutiveSlices.emplace_back(lastDim, currConsecutiveSlices);
     }
 
-    // Mapping from old Slices to new Slices where the range of each old Slice
-    // is a subset of the corresponding new Slice
-    std::unordered_map<SliceNode *, SliceNode *> oldToNewSlices;
-    for (const auto &slices : consecutiveSlices) {
+    // Mapping from old Slices to new Nodes where a Node can either be
+    // i) a merged Slice
+    // ii) a merged Slice + Reshape
+    // iii) a merged Slice + Transpose + Reshape
+    std::unordered_map<SliceNode *, Node *> oldSlicesToNewNodes;
+
+    for (const auto &slicePairs : consecutiveSlices) {
+      auto slicesDim = slicePairs.first;
+      auto slices = slicePairs.second;
+
       if (slices.size() <= 1) {
         continue;
       }
+      if (slicesDim != CN->getDim() &&
+          ((slicesDim != CN->getDim() + 1 && slicesDim != CN->getDim() - 1) ||
+           slices[0]->getResult().dims()[slicesDim] != 1)) {
+        // Optimizations are possible only if:
+        // 1) slices consecutive dimension is the same as concat dimension, or
+        // 2) slices consecutive dimension is adjacent to the concat dimension,
+        //    and the size of each slice along the consecutive dimension is 1.
+        //    NOTE: Checking the size of 0th slice is sufficient, as opposed
+        //    to checking every slice. If the slices can be concatenated,
+        //    each slice size must be equal.
```

Comment:
Just to make this more clear, can we reword it to something like:

NOTE: Checking the slicesDim dimension of 0th slice is sufficient, as opposed to checking every slice. If the slices can be concatenated (and they're being concatenated along a different dimension), then each slicesDim dim must be equal.